baseplate/context/memcache.py

```diff
@@ -0,0 +1,124 @@
+    * ``connect_timeout``: a float representing seconds to wait for a connection to
+        memcached server. Defaults to the underlying socket default timeout.
+    * ``timeout``: a float representing seconds to wait for calls on the
+        socket connected to memcache. Defaults to the underlying socket default
+        timeout.
+
+    """
+
+    assert prefix.endswith(".")
+    config_prefix = prefix[:-1]
+    cfg = config.parse_config(app_config, {
+        config_prefix: {
+            "endpoint": config.Endpoint,
+            "max_pool_size": config.Optional(config.Integer, default=None),
+            "connect_timeout": config.Optional(config.Float, default=None),
+            "timeout": config.Optional(config.Float, default=None),
+        },
+    })
+
+    options = getattr(cfg, config_prefix)
+
+    if options.max_pool_size is not None:
+        kwargs.setdefault("max_pool_size", options.max_pool_size)
+    if options.connect_timeout is not None:
+        kwargs.setdefault("connect_timeout", options.connect_timeout)
+    if options.timeout is not None:
+        kwargs.setdefault("timeout", options.timeout)
+
+    return PooledClient(options.endpoint.address, **kwargs)
+
+
+class MemcacheContextFactory(ContextFactory):
+    """Memcache client context factory.
+
+    This factory will attach a
+    :py:class:`~baseplate.context.memcache.MonitoredMemcacheConnection` to an
+    attribute on the :term:`context object`. When memcache commands are executed
+    via this connection object, they will use connections from the provided
+    `PooledClient <https://pymemcache.readthedocs.io/en/latest/apidoc/pymemcache.client.base.html#pymemcache.client.base.PooledClient>`_
+    and automatically record diagnostic information.
+
+    :param pymemcache.client.base.PooledClient pooled_client: A pooled client.
+
+    :returns: :py:class:`~baseplate.context.memcache.MonitoredMemcacheConnection`
+
+    """
+    def __init__(self, pooled_client):
+        self.pooled_client = pooled_client
+
+    def make_object_for_context(self, name, server_span):
+        return MonitoredMemcacheConnection(name, server_span, self.pooled_client)
+
+
+class MonitoredMemcacheConnection(PooledClient):
+    """Memcache connection that collects diagnostic information.
+
+    This connection acts like a
+    `PooledClient <https://pymemcache.readthedocs.io/en/latest/apidoc/pymemcache.client.base.html#pymemcache.client.base.PooledClient>`_
+    except that operations are wrapped with diagnostic collection. Some methods
+    may not yet be wrapped with monitoring, please request assistance if any
+    needed methods are not being monitored.
+
+    """
+
+    def __init__(self, context_name, server_span, pooled_client):
+        self.context_name = context_name
+        self.server_span = server_span
+        self.pooled_client = pooled_client
+
+    #pylint: disable=no-self-argument
+    def proxy_with_instrumentation(fn_name):
+        def fn(self, *a, **kw):
```

Comment:
sorry, i meant this one, `fn`, since it actually stays in the call stack during regular operation and so would be in tracebacks. can this guy have a more obvious name?